File: solutions/python/longest_valid_parentheses.py
import random
import sys


# @include
def longest_valid_parentheses(s):
    max_length, end, left_parentheses_indices = 0, -1, []
    for i, c in enumerate(s):
        if c == '(':
            left_parentheses_indices.append(i)
        elif not left_parentheses_indices:
            end = i
        else:
            left_parentheses_indices.pop()
            start = (left_parentheses_indices[-1]
                     if left_parentheses_indices else end)
            max_length = max(max_length, i - start)
    return max_length
# @exclude


def longest_valid_parentheses_constant_space(s):
    def parse_from_side(s, paren):
        max_length = num_parens_so_far = length = 0
        for c in s:
            if c == paren:
                num_parens_so_far, length = num_parens_so_far + 1, length + 1
            else:  # c != paren
                if num_parens_so_far <= 0:
                    num_parens_so_far = length = 0
                else:
                    num_parens_so_far -= 1
                    length += 1
                    if num_parens_so_far == 0:
                        max_length = max(max_length, length)
        return max_length

    return max(parse_from_side(s, '('), parse_from_side(reversed(s), ')'))


def small_test():
    assert longest_valid_parentheses(')(((())()(()(') == 6
    assert longest_valid_parentheses('((())()(()(') == 6
    assert longest_valid_parentheses(')(') == 0
    assert longest_valid_parentheses('()') == 2
    assert longest_valid_parentheses('') == 0
    assert longest_valid_parentheses('()()())') == 6
    assert longest_valid_parentheses_constant_space(')(((())()(()(') == 6
    assert longest_valid_parentheses_constant_space('((())()(()(') == 6
    assert longest_valid_parentheses_constant_space(')(') == 0
    assert longest_valid_parentheses_constant_space('()') == 2
    assert longest_valid_parentheses_constant_space('') == 0
    assert longest_valid_parentheses_constant_space('()()())') == 6


def rang_string(length):
    return ''.join('(' if random.randint(0, 1) == 0 else ')'
                   for _ in range(length))


def main():
    small_test()
    if len(sys.argv) == 2:
        s = sys.argv[1]
        print('s =', s)
        print(longest_valid_parentheses(s))
    else:
        for _ in range(1000):
            s = rang_string(random.randint(0, 100000))
            assert longest_valid_parentheses_constant_space(
                s) == longest_valid_parentheses(s)


if __name__ == '__main__':
    main()

File: solutions/java/LongestValidParentheses.java
import java.util.Deque;
import java.util.LinkedList;
import java.util.Random;

public class LongestValidParentheses {
  // @include
  public static int longestValidParentheses(String s) {
    int maxLength = 0, end = -1;
    Deque<Integer> leftParenthesesIndices = new LinkedList<>();
    for (int i = 0; i < s.length(); ++i) {
      if (s.charAt(i) == '(') {
        leftParenthesesIndices.addFirst(i);
      } else if (leftParenthesesIndices.isEmpty()) {
        end = i;
      } else {
        leftParenthesesIndices.removeFirst();
        int start = leftParenthesesIndices.isEmpty()
                        ? end
                        : leftParenthesesIndices.peekFirst();
        maxLength = Math.max(maxLength, i - start);
      }
    }
    return maxLength;
  }
  // @exclude

  private static int longestValidParenthesesConstantSpace(String s) {
    return Math.max(parseFromSide(s, '(', 0, s.length(), 1),
                    parseFromSide(s, ')', s.length() - 1, -1, -1));
  }

  private static int parseFromSide(String s, char paren, int begin, int end,
                                   int dir) {
    int maxLength = 0, numParensSoFar = 0, length = 0;
    for (int i = begin; i != end; i += dir) {
      if (s.charAt(i) == paren) {
        ++numParensSoFar;
        ++length;
      } else { // s.charAt(i) != paren
        if (numParensSoFar <= 0) {
          numParensSoFar = length = 0;
        } else {
          --numParensSoFar;
          ++length;
          if (numParensSoFar == 0) {
            maxLength = Math.max(maxLength, length);
          }
        }
      }
    }
    return maxLength;
  }

  private static void smallTest() {
    assert(longestValidParentheses(")(((())()(()(") == 6);
    assert(longestValidParentheses("((())()(()(") == 6);
    assert(longestValidParentheses(")(") == 0);
    assert(longestValidParentheses("()") == 2);
    assert(longestValidParentheses("") == 0);
    assert(longestValidParentheses("()()())") == 6);
    assert(longestValidParenthesesConstantSpace(")(((())()(()(") == 6);
    assert(longestValidParenthesesConstantSpace("((())()(()(") == 6);
    assert(longestValidParenthesesConstantSpace(")(") == 0);
    assert(longestValidParenthesesConstantSpace("()") == 2);
    assert(longestValidParenthesesConstantSpace("") == 0);
    assert(longestValidParenthesesConstantSpace("()()())") == 6);
  }

  private static String randString(int length) {
    Random r = new Random();
    StringBuilder result = new StringBuilder(length);
    while (length-- > 0) {
      result.append((r.nextInt(1) == 0 ? '(' : ')'));
    }
    return result.toString();
  }

  public static void main(String[] args) {
    smallTest();
    if (args.length == 1) {
      String s = args[0];
      System.out.println("s = " + s);
      System.out.println(longestValidParentheses(s));
    } else {
      Random r = new Random();
      for (int times = 0; times < 1000; ++times) {
        String s = randString(r.nextInt(100000));
        assert(longestValidParenthesesConstantSpace(s)
               == longestValidParentheses(s));
      }
    }
  }
}
